repos/system_upgrade/el7toel8/actors/checkminimumrequirements/libraries/library.py

```diff
@@ -0,0 +1,89 @@
+import os
+import re
+
+from leapp.models import HardwareInfo
+from leapp.libraries.common.config import architecture
+from leapp.libraries.stdlib import api
+from leapp.reporting import create_report
+from leapp import reporting
+
+min_req_data = {
+    'memory': {
+        architecture.ARCH_X86_64: 1536,  # 1.5G
+        architecture.ARCH_ARM64: 2048,  # 2Gb
+        architecture.ARCH_PPC64LE: 2048,  # 2Gb
+        architecture.ARCH_S390X: 1024  # 1Gb
+    },
+    'kernel_version': '3.10'
+}
+
+
+def _get_kernel_version(hardware_info):
+    hardware_info.kernel_version = re.findall(r'\d+\.\d+\.\d+', os.uname()[2])[0]
+
+
+def _check_kernel_version(minimum_req_error, hardware_info):
+    kernel_ver_info = tuple(map(int, hardware_info.kernel_version.split('.')))
+    if kernel_ver_info < (3, 10, 0):
+        minimum_req_error.append('Minimu kernel version: {}    found: {}'.format(
+            min_req_data['kernel_version'], hardware_info.kernel_version))
+
+
+def _get_memory_info(hardware_info):
+    mem_info = dict((i.split()[0].rstrip(':'), int(i.split()[1])) for i in open('/proc/meminfo').readlines())
+    total_mem_kib = mem_info['MemTotal']
+
+    hardware_info.mem_total = total_mem_kib
+
+
+def _check_memory(minimum_req_error, mem_info):
+    if (architecture.matches_architecture(architecture.ARCH_X86_64) and
+            mem_info.mem_total < min_req_data['memory'][architecture.ARCH_X86_64]):
+        minimum_req_error.append(
+            'Total Memory required: {}   found: {}'.format(min_req_data['memory'][architecture.ARCH_X86_64], mem_info))
+    if (architecture.matches_architecture(architecture.ARCH_ARM64)
+            and mem_info.mem_total < min_req_data['memory'][architecture.ARCH_ARM64]):
+        minimum_req_error.append(
+            'Total Memory required: {}    found: {}'.format(min_req_data['memory'][architecture.ARCH_ARM64], mem_info))
+    if (architecture.matches_architecture(architecture.ARCH_PPC64LE) and
+            mem_info.mem_total < min_req_data['memory'][architecture.ARCH_PPC64LE]):
+        minimum_req_error.append(
+            'Total Memory required: {}    found: {}'.format(min_req_data['memory'][architecture.ARCH_PPC64LE],
+                                                            mem_info))
+    if (architecture.matches_architecture(architecture.ARCH_S390X) and
+            mem_info.mem_total < min_req_data['memory'][architecture.ARCH_S390X]):
+        minimum_req_error.append(
+            'Total Memory required: {}    found: {}'.format(min_req_data['memory'][architecture.ARCH_S390X], mem_info))
```

Comment:
It doesn't make sense to produce the message here. Why? Only what I want is information what is required, what is discovered (tuple with two values). The information can be represented somewhere else. Just on the one line of the code. Now I see several duplicit lines for which I do not see any reason.

I think much better solution seems to return the tuple if requirements have not been met, or None if everything is ok - I am not telling that even this is ideal, but I would rather now what's wrong and work with it somehow (see the related sources comment) than just have a msg.

Reply:
Besides what Petr said, _check_memory could use some refactor (deduplication and you can have just 1 arch, right, why check for all 4?)
```
def _check_memory(mem_info):
"""
Returns (True, '') if memory check passed or (False, error_msg) if it didn't 
"""
    for arch, min_req in min_req_data['memory'].iteritems():
        if architecture.matches_architecture(arch):
            is_ok = mem_info.mem_total >= min_req:
            msg = '' if is_ok else 'Total memory required: {} found: {}'.format(min_req, mem_info.mem_total)
            return (is_ok, msg)
```